File: android-app/app/src/main/java/com/sassyconsulting/sassytalkie/PresenceClient.kt
// Copyright (c) 2026 Shane Smith / Sassy Consulting LLC. All rights reserved.
// Proprietary source. This notice is Copyright Management Information (17 U.S.C. 1202); removal or alteration prohibited.
// CodeMark: SCLLC1-sassytalkie-GJTGG7M3YYW5
package com.sassyconsulting.sassytalkie

import android.content.Context
import android.util.Log
import com.google.firebase.messaging.FirebaseMessaging
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Thin client for the relay's /presence endpoint. Lets the app register its
 * FCM token under (current room, this install's peer_id) so the relay can
 * fire a wake push when audio arrives for a peer whose WS has dropped.
 *
 * Stateless / static — fire and forget. Callers should run from a background
 * coroutine on Dispatchers.IO.
 */
object PresenceClient {
    private const val TAG = "PresenceClient"

    private val http: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(15, TimeUnit.SECONDS)
            .build()
    }

    private val JSON_MEDIA = "application/json".toMediaType()

    /** Returns true on 2xx, false otherwise. Logs failures but never throws. */
    fun upload(context: Context, roomId: String, fcmToken: String): Boolean {
        if (roomId.isBlank() || fcmToken.isBlank()) return false
        val peerId = InstallId.get(context)
        val body = JSONObject().apply {
            put("room", roomId)
            put("peer", peerId)
            put("token", fcmToken)
        }
        val req = Request.Builder()
            .url("${SessionShareLink.RELAY_BASE}/presence")
            .post(body.toString().toRequestBody(JSON_MEDIA))
            .build()
        return try {
            http.newCall(req).execute().use { r ->
                if (!r.isSuccessful) {
                    Log.w(TAG, "presence upload failed: HTTP ${r.code}")
                }
                r.isSuccessful
            }
        } catch (t: Throwable) {
            Log.w(TAG, "presence upload exception: ${t.message}")
            false
        }
    }

    /**
     * Fetch the current FCM registration token and upload it to /presence
     * for [roomId]. No-op when FCM hasn't been configured (no
     * google-services.json) — the token fetch will throw and we just log.
     *
     * Fire-and-forget: returns immediately, work happens on FCM's background
     * thread + an OkHttp dispatcher thread.
     */
    fun uploadCurrentToken(context: Context, roomId: String) {
        if (roomId.isBlank()) return
        val appCtx = context.applicationContext
        try {
            FirebaseMessaging.getInstance().token
                .addOnSuccessListener { token ->
                    if (!token.isNullOrBlank()) {
                        Thread {
                            upload(appCtx, roomId, token)
                        }.apply { name = "presence-upload"; isDaemon = true }.start()
                    }
                }
                .addOnFailureListener { e ->
                    Log.w(TAG, "FCM token fetch failed: ${e.message}")
                }
        } catch (t: Throwable) {
            // FirebaseApp not initialized (no google-services.json) — safe skip.
            Log.d(TAG, "FCM not configured, skipping presence upload: ${t.message}")
        }
    }

    /** DELETE the (room, peer) row — call on session clear / sign out. */
    fun remove(context: Context, roomId: String): Boolean {
        if (roomId.isBlank()) return false
        val peerId = InstallId.get(context)
        val body = JSONObject().apply {
            put("room", roomId)
            put("peer", peerId)
        }
        val req = Request.Builder()
            .url("${SessionShareLink.RELAY_BASE}/presence")
            .delete(body.toString().toRequestBody(JSON_MEDIA))
            .build()
        return try {
            http.newCall(req).execute().use { r -> r.isSuccessful }
        } catch (t: Throwable) {
            Log.w(TAG, "presence remove exception: ${t.message}")
            false
        }
    }
}

File: ios-native/PresenceClient.swift
// Copyright (c) 2026 Shane Smith / Sassy Consulting LLC. All rights reserved.
// Proprietary source. This notice is Copyright Management Information (17 U.S.C. 1202); removal or alteration prohibited.
// CodeMark: SCLLC1-sassytalkie-IOSAPNSPRES7K
//
//  PresenceClient.swift
//  Thin client for the relay's /presence endpoint (parity with Android
//  PresenceClient.kt). Registers this install's APNs device token under
//  (room, peer) so the worker can fire a wake push when the WS has dropped.
//
//  Auth: POST/DELETE require Authorization: Bearer <room capability> minted
//  from /auth?room=&peer= (peer-bound — see presence.js verifyCapabilityIdentity).
//  Never POST an empty or invented token; wait for a real APNs device token.

import Foundation
import UIKit

enum PresenceClient {
    static let relayBase = "https://relay.sassyconsultingllc.com"
    private static let tokenDefaultsKey = "apns_device_token_hex"
    private static let peerDefaultsKey = "install_peer_id"

    /// Stable per-install peer id — must match RelayClient's WS `peer=` query.
    static var peerId: String {
        if let existing = UserDefaults.standard.string(forKey: peerDefaultsKey), !existing.isEmpty {
            return existing
        }
        let fresh = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
        UserDefaults.standard.set(fresh, forKey: peerDefaultsKey)
        return fresh
    }

    /// Hex-encoded APNs device token, or nil until registration succeeds.
    static var storedDeviceToken: String? {
        get {
            let s = UserDefaults.standard.string(forKey: tokenDefaultsKey)
            return (s?.isEmpty == false) ? s : nil
        }
        set {
            if let newValue, !newValue.isEmpty {
                UserDefaults.standard.set(newValue, forKey: tokenDefaultsKey)
            } else {
                UserDefaults.standard.removeObject(forKey: tokenDefaultsKey)
            }
        }
    }

    static func rememberDeviceToken(_ data: Data) {
        let hex = data.map { String(format: "%02x", $0) }.joined()
        guard !hex.isEmpty else { return }
        storedDeviceToken = hex
    }

    /// Fire-and-forget: upload the stored APNs token for [roomId] when one exists.
    static func uploadCurrentToken(roomId: String) {
        guard !roomId.isEmpty, let token = storedDeviceToken, !token.isEmpty else { return }
        DispatchQueue.global(qos: .utility).async {
            _ = upload(roomId: roomId, deviceToken: token)
        }
    }

    /// Returns true on 2xx. Never throws; never logs the token value.
    @discardableResult
    static func upload(roomId: String, deviceToken: String) -> Bool {
        guard !roomId.isEmpty, !deviceToken.isEmpty else { return false }
        let peer = peerId
        guard let cap = fetchCapabilityToken(room: roomId, peer: peer) else { return false }
        let body: [String: String] = [
            "room": roomId,
            "peer": peer,
            "token": deviceToken,
            "platform": "apns",
        ]
        guard let url = URL(string: "\(relayBase)/presence"),
              let data = try? JSONSerialization.data(withJSONObject: body) else { return false }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(cap)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = data
        return syncBool(req)
    }

    /// DELETE the (room, peer) row — call on session wipe.
    @discardableResult
    static func remove(roomId: String) -> Bool {
        guard !roomId.isEmpty else { return false }
        let peer = peerId
        guard let cap = fetchCapabilityToken(room: roomId, peer: peer) else { return false }
        let body: [String: String] = ["room": roomId, "peer": peer]
        guard let url = URL(string: "\(relayBase)/presence"),
              let data = try? JSONSerialization.data(withJSONObject: body) else { return false }
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        req.setValue("Bearer \(cap)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = data
        return syncBool(req)
    }

    private static func fetchCapabilityToken(room: String, peer: String) -> String? {
        var comps = URLComponents(string: "\(relayBase)/auth")
        comps?.queryItems = [
            URLQueryItem(name: "room", value: room),
            URLQueryItem(name: "peer", value: peer),
        ]
        guard let url = comps?.url else { return nil }
        let sem = DispatchSemaphore(value: 0)
        var token: String?
        PinnedURLSession.shared.dataTask(with: url) { data, response, _ in
            defer { sem.signal() }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard (200..<300).contains(code),
                  let data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let t = json["token"] as? String, !t.isEmpty else {
                NSLog("SassyTalkie presence: auth failed HTTP %d", code)
                return
            }
            token = t
        }.resume()
        _ = sem.wait(timeout: .now() + 15)
        return token
    }

    private static func syncBool(_ req: URLRequest) -> Bool {
        let sem = DispatchSemaphore(value: 0)
        var ok = false
        PinnedURLSession.shared.dataTask(with: req) { _, response, _ in
            defer { sem.signal() }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            ok = (200..<300).contains(code)
            if !ok {
                NSLog("SassyTalkie presence: HTTP %d", code)
            }
        }.resume()
        _ = sem.wait(timeout: .now() + 15)
        return ok
    }
}

extension Notification.Name {
    /// Posted when a wake push (kind/type=wake) arrives or is tapped.
    static let sassyTalkieWake = Notification.Name("sassyTalkieWake")
}
